fail2ban/server/ipdns.py

```diff
@@ -0,0 +1,423 @@
+		"""
+		self._family = socket.AF_UNSPEC
+		self._addr = 0
+		self._plen = 0
+		self._maskplen = None
+		self._raw = ""
+
+		for family in [socket.AF_INET, socket.AF_INET6]:
+			try:
+				binary = socket.inet_pton(family, ipstr)
+				self._family = family
+				break
+			except socket.error:
+				continue
+
+		if self._family == socket.AF_INET:
+			# convert host to network byte order
+			self._addr, = struct.unpack("!L", binary)
+			self._plen = 32
+
+			# mask out host portion if prefix length is supplied
+			if cidr is not None and cidr >= 0:
+				mask = ~(0xFFFFFFFFL >> cidr)
+				self._addr &= mask
+				self._plen = cidr
+
+		elif self._family == socket.AF_INET6:
+			# convert host to network byte order
+			hi, lo = struct.unpack("!QQ", binary)
+			self._addr = (hi << 64) | lo
+			self._plen = 128
+
+			# mask out host portion if prefix length is supplied
+			if cidr is not None and cidr >= 0:
+				mask = ~(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFL >> cidr)
+				self._addr &= mask
+				self._plen = cidr
+
+			# if IPv6 address is a IPv4-compatible, make instance a IPv4
+			elif self.isInNet(IPAddr.IP6_4COMPAT):
+				self._addr = lo & 0xFFFFFFFFL
+				self._family = socket.AF_INET
+				self._plen = 32
+		else:
+			# string couldn't be converted neither to a IPv4 nor
+			# to a IPv6 address - retain raw input for later use
+			# (e.g. DNS resolution)
+			self._raw = ipstr
+
+	def __repr__(self):
+		return self.ntoa
+
+	def __str__(self):
+		return self.ntoa
+	
+	@property
+	def addr(self):
+		return self._addr
+
+	@property
+	def family(self):
+		return self._family
+
+	@property
+	def plen(self):
+		return self._plen
+
+	@property
+	def raw(self):
+		"""The raw address
+
+		Should only be set to a non-empty string if prior address
+		conversion wasn't possible
+		"""
+		return self._raw
+
+	@property
+	def isValid(self):
+		"""Either the object corresponds to a valid IP address
+		"""
+		return self._family != socket.AF_UNSPEC
+
+	def __eq__(self, other):
+		if not isinstance(other, IPAddr):
+			if other is None: return False
+			other = IPAddr(other)
+		if self._family != other._family: return False
+		if self._family == socket.AF_UNSPEC:
+			return self._raw == other._raw
+		return (
+			(self._addr == other._addr) and
+			(self._plen == other._plen)
+		)
+
+	def __ne__(self, other):
+		return not (self == other)
+
+	def __lt__(self, other):
+		if not isinstance(other, IPAddr):
+			if other is None: return False
+			other = IPAddr(other)
+		return self._family < other._family or self._addr < other._addr
+
+	def __add__(self, other):
+		if not isinstance(other, IPAddr):
+			other = IPAddr(other)
+		return "%s%s" % (self, other)
+
+	def __radd__(self, other):
+		if not isinstance(other, IPAddr):
+			other = IPAddr(other)
+		return "%s%s" % (other, self)
+
+	def __hash__(self):
+		# should be the same as by string (because of possible compare with string):
+		return hash(self.ntoa)
+		#return hash(self._addr)^hash((self._plen<<16)|self._family)
```

Comment:
That is exactly one of the many things that, I call "implicit string converting" :)
This allows preventing of many modifications in other modules around. Just try to comment this and uncomment another, and starts test-cases - you will see how many tests will fail.
We would anyway rewrite a lazy ntoa (with direct string representation), because for 99% of all IPAddr (with cidr = -1) it will be exact the same as its `instr` in constructor (resp. __new__).